From 5b8f7cdc4ae4413bdfd581a28623192ffd9e1e93 Mon Sep 17 00:00:00 2001
From: florent pirou <florent.pirou@intel.com>
Date: Wed, 25 Mar 2026 19:12:24 -0700
Subject: [PATCH 09/80] ipu-acpi: fallback for camera pdata without rotation
 param

```
[   45.223993] IPU ACPI:  d4xx current i2c_bdf 0000:00:15.1 serdes 0-1, prior 0000:00:15.1
[   45.228600] IPU ACPI:  d4xx current i2c_addr 0x14 serdes 0-1, prior 0x12
[   45.228602] IPU ACPI:  d4xx current i2c_addr 0x48 sensor 0-1, prior 0x48
[   45.228603] IPU ACPI: Increment d4xx namespacing on aggregated-link sensors 0
```
* ipu-acpi: fallback for camera pdata with wrong d4xx pprval and pprnum entries

Signed-off-by: florent pirou <florent.pirou@intel.com>
---
 drivers/media/platform/intel/ipu-acpi-pdata.c | 34 +++++++++++++------
 1 file changed, 24 insertions(+), 10 deletions(-)

diff --git a/drivers/media/platform/intel/ipu-acpi-pdata.c b/drivers/media/platform/intel/ipu-acpi-pdata.c
index d64f1bff..6c9a03dd 100644
--- a/drivers/media/platform/intel/ipu-acpi-pdata.c
+++ b/drivers/media/platform/intel/ipu-acpi-pdata.c
@@ -832,6 +832,17 @@ int set_pdata(struct ipu_isys_subdev_info **sensor_sd,
 			pr_err("IPU ACPI: Invalid MIPI Port : %d", port);
 
 		pdata->link_freq_mbps = link_freq;
+#if IS_ENABLED(CONFIG_VIDEO_D4XX)
+			if ((!strcmp(sensor_name, D457_NAME)) &&
+			     (lanes != 2 ||
+			      deser_lanes == 0 ||
+			      subdev_num != 1)) {
+			  //fallback from PDATA misconfiguration corner-cases that leads d4xx segfault 
+				deser_lanes = 2;
+				lanes = 2;
+				subdev_num = 1;
+			}
+#endif
 		pdata->bus_type = (*sensor_sd)->csi2->bus_type;
 		pdata->deser_nlanes = deser_lanes;
 		pdata->ser_nlanes = lanes;
@@ -868,28 +879,31 @@ void set_serdes_info(struct device *dev, const char *sensor_name,
 	serdes_info.i2c_num = cam_data->i2c_num;
 
 	i = 1;
+
+#if IS_ENABLED(CONFIG_VIDEO_D4XX)
+	/* prior serializer mapped addr */
+	serdes_info.ser_map_addr_2 = serdes_info.ser_map_addr;
+#endif
 	/* serializer mapped addr */
 	serdes_info.ser_map_addr = cam_data->i2c[i++].addr;
 	/* serializer physical addr */
 	serdes_info.ser_phys_addr = ser_physical_addr;
+
+#if IS_ENABLED(CONFIG_VIDEO_D4XX)
+	/* prior serializer mapped addr */
+	serdes_info.sensor_map_addr_2 = serdes_info.sensor_map_addr;
+#endif
 	/* sensor mapped addr */
 	serdes_info.sensor_map_addr = cam_data->i2c[i++].addr;
 
 #if IS_ENABLED(CONFIG_VIDEO_D4XX)
-	/* serializer mapped addr */
-	if ( serdes_info.deser_num > 0 )
-		serdes_info.ser_map_addr_2 = serdes_info.ser_map_addr;
-	serdes_info.ser_map_addr = cam_data->i2c[cam_data->i2c_num - 2].addr;
-
 	/* deserializer mapped addr */
-	if ( serdes_info.deser_num > 0 )
-		serdes_info.des_map_addr_2 = serdes_info.des_map_addr;
+	serdes_info.des_map_addr_2 = serdes_info.des_map_addr;
 	serdes_info.des_map_addr = cam_data->i2c[cam_data->i2c_num - 3].addr;
 
 	/* serdes mapped i2c-adapter */
-	if ( serdes_info.deser_num > 0 )
-		strscpy(&serdes_info.i2c_adapter_bdf_2[0], serdes_info.i2c_adapter_bdf,
-			sizeof(serdes_info.i2c_adapter_bdf_2));
+	strscpy(&serdes_info.i2c_adapter_bdf_2[0], serdes_info.i2c_adapter_bdf,
+		sizeof(serdes_info.i2c_adapter_bdf_2));
 	strscpy(&serdes_info.i2c_adapter_bdf[0], cam_data->i2c[cam_data->i2c_num - 1].bdf,
 		sizeof(serdes_info.i2c_adapter_bdf));
 #endif
-- 
2.25.1

